onionreq/signal-xed25519.cpp

```diff
@@ -2,46 +2,141 @@
 
 #include "signal-xed25519.hpp"
 #include <cassert>
+#include <cstring>
 #include <stdexcept>
 #include <sodium/randombytes.h>
+#include <sodium/crypto_internal_fe25519.h>
+#include <sodium/crypto_sign_ed25519.h>
+#include <sodium/crypto_scalarmult_ed25519.h>
+#include <sodium/crypto_core_ed25519.h>
+#include <sodium/crypto_generichash_blake2b.h>
 
-extern "C" {
-#include <xeddsa.h>
-#include <crypto_additions.h>
+namespace onionreq::xed25519 {
+
+
+template <size_t N> using bytes = std::array<unsigned char, N>;
+
+namespace {
+
+// constant time `if (b) f = g;` implementation
+template <size_t N>
+void constant_time_conditional_assign(bytes<N>& f, const bytes<N>& g, bool b) {
+    bytes<N> x;
+    for (size_t i = 0; i < x.size(); i++)
+        x[i] = f[i] ^ g[i];
+    unsigned char mask = (unsigned char) (-(signed char) b);
+    for (size_t i = 0; i < x.size(); i++)
+        x[i] &= mask;
+    for (size_t i = 0; i < x.size(); i++)
+        f[i] ^= x[i];
 }
 
-namespace onionreq::xed25519 {
+void fe25519_montx_to_edy(fe25519 y, const fe25519 u) {
```

Comment:
`fe25519 & y` ?

Reply:
fe25519 is typedef'ed to either a uint64_t[5] or a uint32_t[10], depending on the architecture, but both get passed as pointers.

Reply:
ew